ksql-engine/src/main/java/io/confluent/ksql/query/id/HybridQueryIdGenerator.java

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright 2019 Confluent Inc.
+ *
+ * Licensed under the Confluent Community License (the "License"); you may not use
+ * this file except in compliance with the License.  You may obtain a copy of the
+ * License at
+ *
+ * http://www.confluent.io/confluent-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.query.id;
+
+import com.google.common.annotations.VisibleForTesting;
+
+/**
+ * Contains both a Sequential Query Id Generator and a Specific QueryId Generator
+ * Can toggle between the two depending on which one is currently activated
+ * The query id generation method was changed from using
+ */
+public class HybridQueryIdGenerator implements QueryIdGenerator {
+
+  private final SequentialQueryIdGenerator legacyGenerator;
+  private final SpecificQueryIdGenerator newGenerator;
+  private QueryIdGenerator activeGenerator;
+
+  public HybridQueryIdGenerator() {
+    this(0L);
+  }
+
+  public HybridQueryIdGenerator(final long initialValue) {
+    this.legacyGenerator = new SequentialQueryIdGenerator(initialValue);
+    this.newGenerator = new SpecificQueryIdGenerator();
+    this.activeGenerator = this.legacyGenerator;
+  }
+
+  @VisibleForTesting
+  HybridQueryIdGenerator(
+      final SequentialQueryIdGenerator sequentialQueryIdGenerator,
+      final SpecificQueryIdGenerator specificQueryIdGenerator
+  ) {
+    this.legacyGenerator = sequentialQueryIdGenerator;
+    this.newGenerator = specificQueryIdGenerator;
+    this.activeGenerator = this.legacyGenerator;
+  }
+
+  public void activateNewGenerator(final long nextId) {
+    newGenerator.setNextId(nextId);
+    this.activeGenerator = newGenerator;
+  }
+
+  public void activateLegacyGenerator() {
+    activeGenerator = legacyGenerator;
+  }
+
+  @Override
+  public long peekNext() {
+    return activeGenerator.peekNext();
+  }
+
+  @Override
+  public String getNext() {
+    return activeGenerator.getNext();
+  }
+
+  @Override
+  public QueryIdGenerator createSandbox() {
+    return new SequentialQueryIdGenerator(activeGenerator.peekNext());
```

Comment:
I think we should create a hybrid generator with sandboxes for both - otherwise we run the risk of generating duplicate queryIDs while validating queries (or validating against a different queryID than we actually are planning to issue)

Reply:
That could be doable. So would we want the sandboxed HybridGenerator to also have the active generator be set to whichever was active at the time of creation? If its in SpecificIdGenerator (newGenerator) mode, that might be a bit tricky to get it working when validating.

setNextId() would need to be called in the RequestValidator to keep updating the sandbox generator in case there are multiple queries being created in one request from the user. We could grab the previous offset and increment it based on the sequence of statements being validated currently (if it's the 3rd statement, we call setNextId(previous offset + 3). I would also need a way to expose the sandbox generator from EngineContext as I have no way to reference it right now from RequestValidator. https://github.com/confluentinc/ksql/pull/3354/files#diff-78b83625c0f213c7b6dc1e09befdcb21R112

Reply:
If we take the comment above that generating a sandbox for `SpecificIdGenerator` just creates a sequential one, then we have no problem (e.g. it doesn't matter what the active generator type is, we just always call `.createSandbox()` on it and let it handle the details)